### PR TITLE
add class SocketMaster

### PR DESCRIPTION
**Pull request by deanlee**
Manage Context,Poller, PubSocket and Subsocket  in one class. delete them in the right order.

### CODE REVIEW

File: selfdrive/common/socketmaster.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  std::vector<Message *> pollMessages(int timeout) {
+    assert(poller != NULL);
+    std::vector<Message *> messages;
+    while (true) {
+      auto polls = poller->poll(timeout);
+      if (polls.size() == 0) {
+        break;
+      }
+      for (auto sock : polls) {
+        Message *msg = sock->receive();
```
Comment (quillford) [inline selfdrive/common/socketmaster.h:48]: I think this should be a non blocking, we're already waiting on the poller timeout

**Comment by quillford** [comment]
I think we should call this SubMaster like the python version, and this should go with the rest of messaging in cereal/messaging.

**Comment by deanlee** [comment]
I thought about that before. but it looks like that there are no much thing to do with PubSocket,we just need to  guaranteed it will be deleted. And it will introduce the complexity of managing context if we 
split into two classes.

**Comment by pd0wm** [comment]
I also prefer a separate SubMaster/PubMaster class. Could be helpful when doing dependency injection for testing.

Can you submit a PR for cereal? I'd also like the API to SubMaster to be comparable to the python version (with the alive and valid checks), we should really have the same checks in the c code.

**Comment by deanlee** [comment]
ok i'll try

**Pull request closed without merging**